### PR TITLE
 [BEAM-7866] Fix python ReadFromMongoDB potential data loss issue

### PR DESCRIPTION
**Pull request by y1chi**
Change the liquid sharding split strategy to use splitVector command in mongodb to make sure the split range is deterministic and non-overlapping.
Remove all IO operations in initializing the class to avoid slow startup or unavailability issues.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by y1chi** [comment]
Python MongoDBIO Integration Test

**Comment by y1chi** [comment]
Run Python MongoDBIO Integration Test

**Comment by y1chi** [comment]
Run Python MongoDBIO_IT

**Comment by y1chi** [comment]
Run Python MongoDBIO_IT

**Comment by y1chi** [comment]
R: @chamikaramj @jkff

File: sdks/python/apache_beam/io/mongodbio.py [hunk 1]
```diff
@@ -139,50 +143,77 @@ def __init__(self,
...
+    if bundle_start < stop_position:
+      yield iobase.SourceBundle(weight=desired_bundle_size_in_mb,
+                                source=self,
+                                start_position=bundle_start,
+                                stop_position=stop_position)
 
   def get_range_tracker(self, start_position, stop_position):
     if start_position is None:
-      start_position = 0
+      epoch = datetime.datetime(1970, 1, 1)
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio.py]: When can start_position and stop_position be None in get_range_tracker?

File: sdks/python/apache_beam/io/mongodbio.py [hunk 2]
```diff
@@ -139,50 +143,77 @@ def __init__(self,
...
-      docs = client[self.db][self.coll].find(
-          filter=self.filter, projection=self.projection
-      )[range_tracker.start_position():range_tracker.stop_position()]
-      for index in range(range_tracker.start_position(),
-                         range_tracker.stop_position()):
-        if not range_tracker.try_claim(index):
+      all_filters = self.filter
+      all_filters.update({
+          '_id': {
+              '$gte': range_tracker.start_position(),
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio.py]: What if the original filter already has gte/lt conditions on _id? I think it's fine to prohibit it (at construction time); but you can also take it into account and infer the start/stop position in split() from it.

File: sdks/python/apache_beam/io/mongodbio.py [hunk 3]
```diff
@@ -194,18 +225,72 @@ def display_data(self):
     res['mongo_client_spec'] = self.spec
     return res
 
-  def _get_avg_document_size(self):
+  def _get_split_keys(self, desired_chunk_size, start_pos, end_pos):
+    # if desired chunk size smaller than 1mb, use mongodb default split size of
+    # 1mb
+    if desired_chunk_size < 1:
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio.py]: What are the units of desired_chunk_size? The comment implies mb?

File: sdks/python/apache_beam/io/mongodbio.py [hunk 4]
```diff
@@ -194,18 +225,72 @@ def display_data(self):
...
+
+class _ObjectIdRangeTracker(OrderedPositionRangeTracker):
+  """RangeTracker for tracking mongodb _id of bson ObjectId type."""
+
+  def _id_to_int(self, id):
+    # id object is bytes type with size of 12
+    ints = struct.unpack('>III', id.binary)
+    return 2**64 * ints[0] + 2**32 * ints[1] + ints[2]
+
+  def _int_to_id(self, numbers):
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio.py]: This kind of bignum arithmetic is extremely hard to get right, please make sure the unit tests cover enough corner cases. I would recommend to add some randomized property testing.

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 5]
```diff
@@ -30,38 +32,102 @@
...
 
+class _MockMongoColl(object):
+  def __init__(self, docs):
+    self.docs = docs
+
+  def _filter(self, filter):
+    match = []
+    if not filter:
+      return self
+    start = filter['_id'].get('$gte')
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio_test.py]: Please assert that start, end are both not None

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 6]
```diff
@@ -30,38 +32,102 @@
...
-  @mock.patch('apache_beam.io.mongodbio._BoundedMongoSource'
-              '._get_document_count')
-  @mock.patch('apache_beam.io.mongodbio._BoundedMongoSource'
-              '._get_avg_document_size')
-  def setUp(self, mock_size, mock_count):
-    mock_size.return_value = 10
-    mock_count.return_value = 5
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def setUp(self, mock_client):
+    mock_client.return_value.__enter__.return_value.__getitem__ \
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio_test.py]: It's hard to follow what this line is doing, could you add a comment? Ditto for other similar statements.
I'm wondering if mocking is even the right approach here. Maybe create a fake instead? Seems like that could be much more readable and less fragile w.r.t. precise order of method calls.

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 7]
```diff
@@ -30,38 +32,102 @@
...
+        for i in range(5)
+    ]
+    self._docs = [{'_id': self._ids[i], 'x': i} for i in range(len(self._ids))]
+
     self.mongo_source = _BoundedMongoSource('mongodb://test', 'testdb',
                                             'testcoll')
 
-  def test_estimate_size(self):
-    self.assertEqual(self.mongo_source.estimate_size(), 50)
+  def get_split(self, command, ns, min, max, maxChunkSize, **kwargs):
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio_test.py]: * What does this do? It's not quite clear from the method body
* maxChunkSize should probably be called max_chunk_size

File: sdks/python/apache_beam/io/mongodbio.py [hunk 8]
```diff
@@ -139,50 +143,77 @@ def __init__(self,
...
+      if size is None or size <= 0:
+        raise ValueError('Collection %s not found or total doc size is '
+                         'incorrect' % self.coll)
+      return size
 
   def split(self, desired_bundle_size, start_position=None, stop_position=None):
     # use document cursor index as the start and stop positions
     if start_position is None:
-      start_position = 0
+      epoch = datetime.datetime(1970, 1, 1)
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: makes sense, thanks for the suggest.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 1 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: My understanding is when the first split happens.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 2 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: Added the filter merge logic.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 3 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: yes in Mb. renamed the argument for clarity.

File: sdks/python/apache_beam/io/mongodbio_test.py (hunk 7 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: maxChunkSize is the actual argument name mango client needed(maybe by mistake) so I'm not able to change it to snake case.

File: sdks/python/apache_beam/io/mongodbio_test.py (hunk 6 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: make sense.

**Comment by jkff** [comment]
Thank you for the updates! I'd like to follow my usual review strategy with IO changes and hand off the next review round to @chamikaramj , I'll do the round after that.

File: sdks/python/apache_beam/io/mongodbio.py [hunk 9]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+      return (client[self.db].command(
+          'splitVector',
+          name_space,
+          keyPattern={'_id': 1},
+          min={'_id': start_pos},
+          max={'_id': end_pos},
+          maxChunkSize=desired_chunk_size_in_mb)['splitKeys'])
+
+  def _merge_id_filter(self, range_tracker):
+    all_filters = self.filter.copy()
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: Looks like self.filter can be None (default) ?

File: sdks/python/apache_beam/io/mongodbio.py [hunk 10]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+    if '_id' in all_filters:
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: What does key '_id' contain ? Please add a comment.

File: sdks/python/apache_beam/io/mongodbio.py [hunk 11]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+      id_filter = all_filters['_id']
+      id_filter['$gte'] = (
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: What does key '$gte' contain ? Please add a comment.  (same for other special keys used here).

File: sdks/python/apache_beam/io/mongodbio.py [hunk 12]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+  def _get_head_document_id(self, sort_order):
     with MongoClient(self.uri, **self.spec) as client:
-      return max(client[self.db][self.coll].count_documents(self.filter), 0)
+      cursor = client[self.db][self.coll].find(filter={}, projection=[]).sort([
+          ('_id', sort_order)
+      ]).limit(1)
+      try:
+        return cursor[0]['_id']
+      except IndexError:
+        raise ValueError('Empty Mongodb collection')
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: Or it just didn't contain key '_id' for some reason ?

File: sdks/python/apache_beam/io/mongodbio.py [hunk 13]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+
+
+class _ObjectIdHelper(object):
+  """A Utility class to bson object ids."""
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: "...to manipulate bson object ids."

File: sdks/python/apache_beam/io/mongodbio.py [hunk 14]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+
+  @classmethod
+  def id_to_int(cls, id):
+    # converts object id binary to integer
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: Please add a proper doc comment including variables if possible. (even though containing class is private, better to properly describe these functions that contain non-trivial byte manipulations).

File: sdks/python/apache_beam/io/mongodbio.py [hunk 15]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+    ints = struct.unpack('>III', id.binary)
+    return (ints[0] << 64) + (ints[1] << 32) + ints[2]
+
+  @classmethod
+  def int_to_id(cls, number):
+    # converts integer value to object id. Int value should be less than
+    # (2 ^ 96) so it can be convert to 12 bytes required by object id.
+    if number < 0 or number >= (1 << 96):
+      raise ValueError('number value must be within [0, %s)' % (1 << 96))
+    ints = [(number & 0xffffffff0000000000000000) >> 64,
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: Please make sure that these functions are extensively tested using unit tests.

File: sdks/python/apache_beam/io/mongodbio.py [hunk 16]
```diff
@@ -194,18 +212,110 @@ def display_data(self):
...
+    bytes = struct.pack('>III', *ints)
+    return objectid.ObjectId(bytes)
+
+  @classmethod
+  def increment_id(cls, object_id, inc):
+    # increment object_id binary value by inc value and return new object id.
+    id_number = _ObjectIdHelper.id_to_int(object_id)
+    new_number = id_number + inc
+    if new_number < 0 or new_number >= (1 << 96):
+      raise ValueError('invalid incremental, inc value must be within ['
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: Seems like we already do this validation inside 'int_to_id' function invoked below ?

File: sdks/python/apache_beam/io/mongodbio.py [hunk 17]
```diff
@@ -139,50 +143,64 @@ def __init__(self,
...
-      start_position = 0
+      start_position = self._get_head_document_id(ASCENDING)
     if stop_position is None:
-      stop_position = self.doc_count
-    return OffsetRangeTracker(start_position, stop_position)
+      last_doc_id = self._get_head_document_id(DESCENDING)
+      # increment last doc id binary value by 1 to make sure the last document
+      # is not excluded
+      stop_position = _ObjectIdHelper.increment_id(last_doc_id, 1)
+    return _ObjectIdRangeTracker(start_position, stop_position)
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio.py]: Are object IDs are directly comparable (as required by OrderedPositionRangeTracker) ?

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 18]
```diff
@@ -30,38 +34,136 @@
...
+  def command(self, command, *args, **kwargs):
+    if command == 'collstats':
+      return {'size': 5, 'avgSize': 1}
+    elif command == 'splitVector':
+      return self.get_split_key(command, *args, **kwargs)
+
+  def get_split_key(self, command, ns, min, max, maxChunkSize, **kwargs):
+    # simulate mongo db splitVector command, return split keys base on chunk
+    # size, assuming every doc is of size 1mb
+    start_id = min['_id']
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio_test.py]: Prob. use different variable names since these override system functions.

File: sdks/python/apache_beam/io/mongodbio.py [hunk 19]
```diff
@@ -139,50 +143,64 @@ def __init__(self,
...
+    desired_bundle_size_in_mb = desired_bundle_size // 1024 // 1024
+    split_keys = self._get_split_keys(desired_bundle_size_in_mb, start_position,
+                                      stop_position)
 
     bundle_start = start_position
-    while bundle_start < stop_position:
-      bundle_end = min(stop_position, bundle_start + desired_bundle_count)
-      yield iobase.SourceBundle(weight=bundle_end - bundle_start,
+    for split_key_id in split_keys:
+      if bundle_start is not None or bundle_start >= stop_position:
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: thought I fixed this earlier, weird.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 17 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: yes

File: sdks/python/apache_beam/io/mongodbio_test.py (hunk 18 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: these are the argument key required by mongo client

File: sdks/python/apache_beam/io/mongodbio.py [hunk 20]
```diff
@@ -139,50 +143,64 @@ def __init__(self,
...
-    self.doc_count = self._get_document_count()
-    self.avg_doc_size = self._get_avg_document_size()
-    self.client = None
 
   def estimate_size(self):
-    return self.avg_doc_size * self.doc_count
+    with MongoClient(self.uri, **self.spec) as client:
+      size = client[self.db].command('collstats', self.coll).get('size')
+      if size is None or size <= 0:
+        raise ValueError('Collection %s not found or total doc size is '
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: normally means the collection is invalid or empty.

File: sdks/python/apache_beam/io/mongodbio.py [hunk 21]
```diff
@@ -139,50 +143,64 @@ def __init__(self,
...
+      if bundle_start is not None or bundle_start >= stop_position:
+        break
+      bundle_end = min(stop_position, split_key_id)
+      yield iobase.SourceBundle(weight=desired_bundle_size_in_mb,
                                 source=self,
                                 start_position=bundle_start,
                                 stop_position=bundle_end)
       bundle_start = bundle_end
+    # add range of last split_key to stop_position
+    if bundle_start < stop_position:
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: yes

File: sdks/python/apache_beam/io/mongodbio.py [hunk 22]
```diff
@@ -139,50 +143,64 @@ def __init__(self,
...
+      yield iobase.SourceBundle(weight=desired_bundle_size_in_mb,
+                                source=self,
+                                start_position=bundle_start,
+                                stop_position=stop_position)
 
   def get_range_tracker(self, start_position, stop_position):
     if start_position is None:
-      start_position = 0
+      start_position = self._get_head_document_id(ASCENDING)
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: is there a situation where one of the position is none and the other isn't?

File: sdks/python/apache_beam/io/mongodbio.py (hunk 9 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: it will be set to {} by constructor.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 12 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: it's actually required primary key for every document.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 16 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: This shows the acceptable inc argument range clearer but I can also remove this check.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: moved to utility function

File: sdks/python/apache_beam/io/mongodbio.py [hunk 23]
```diff
@@ -194,18 +197,145 @@ def display_data(self):
...
+          max={'_id': end_pos},
+          maxChunkSize=desired_chunk_size_in_mb)['splitKeys'])
+
+  def _merge_id_filter(self, range_tracker):
+    # Merge the default filter with refined _id field range of range_tracker.
+    all_filters = self.filter.copy()
+
+    # if there are no additional filters, initialize empty additional filters,
+    # see more at https://docs.mongodb.com/manual/reference/operator/query/and/
+    if '$and' not in all_filters:
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio.py]: What if it has one of the other query operators, such as $or $not $nor? Seems much simpler to just construct an $and of self.filter and {id: gte, lt} - or will that error out if self.filter also involves id in any capacity?

File: sdks/python/apache_beam/io/mongodbio.py [hunk 24]
```diff
@@ -194,18 +197,145 @@ def display_data(self):
...
+    # converts object id binary to integer
+    # id object is bytes type with size of 12
+    ints = struct.unpack('>III', id.binary)
+    return (ints[0] << 64) + (ints[1] << 32) + ints[2]
+
+  @classmethod
+  def int_to_id(cls, number):
+    """
+    Args:
+      number(int): The integer value to be used to convert to ObjectId.
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio.py]: Do all versions of Python that this code can run on use arbitrary-precision integers by default?

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 25]
```diff
@@ -30,38 +34,138 @@
...
+  def __getitem__(self, coll_name):
+    return _MockMongoColl(self.docs)
+
+  def command(self, command, *args, **kwargs):
+    if command == 'collstats':
+      return {'size': 5, 'avgSize': 1}
+    elif command == 'splitVector':
+      return self.get_split_key(command, *args, **kwargs)
+
+  def get_split_key(self, command, ns, min, max, maxChunkSize, **kwargs):
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio_test.py]: should be get_split_keys

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 26]
```diff
@@ -30,38 +34,138 @@
...
+        start_index += 1
+      if doc['_id'] <= end_id:
+        end_index += 1
+      else:
+        break
+    # return ids of elements in the range with chunk size skip and exclude
+    # head element.
+    return {
+        'splitKeys':
+        [x['_id'] for x in self.docs[start_index:end_index:maxChunkSize]][1:]
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio_test.py]: Looks like in this method we interpret maxChunkSize as size of the chunk in terms of number of documents - but the real Mongo uses number of megabytes. Is this a problem for this test? If no, it needs to be explained.

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 27]
```diff
@@ -72,72 +176,47 @@ def test_split(self, mock_client):
...
+    self.assertIsInstance(self.mongo_source.get_range_tracker(None, None),
+                          _ObjectIdRangeTracker)
 
   @mock.patch('apache_beam.io.mongodbio.MongoClient')
   def test_read(self, mock_client):
     mock_tracker = mock.MagicMock()
-    mock_tracker.try_claim.return_value = True
-    mock_tracker.start_position.return_value = 0
-    mock_tracker.stop_position.return_value = 2
+    mock_tracker.start_position.return_value = self._ids[0]
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio_test.py]: Please also add a test that uses a tracker whose position boundaries are between object ids, and tests where 1) start_position is less than any existing id 2) end_position is larger than any existing id 3) there are no objects between start and end.

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 28]
```diff
@@ -210,6 +289,47 @@ def test_write_to_mongodb_with_generated_id(self, mock_client):
...
+         objectid.ObjectId('00000000ffffffffffffffff')),
+    ]
+    for (first, second) in test_cases:
+      self.assertEqual(second, _ObjectIdHelper.increment_id(first, -1))
+      self.assertEqual(first, _ObjectIdHelper.increment_id(second, 1))
+
+    for _ in range(100):
+      id = objectid.ObjectId()
+      self.assertLess(id, _ObjectIdHelper.increment_id(id, 1))
+      self.assertGreater(id, _ObjectIdHelper.increment_id(id, -1))
```
Comment (jkff) [inline sdks/python/apache_beam/io/mongodbio_test.py]: Please add tests for position_to_fraction and vice versa. These are by far the easiest to get wrong.

File: sdks/python/apache_beam/io/mongodbio.py [hunk 29]
```diff
@@ -194,18 +197,145 @@ def display_data(self):
...
-            'Collection %s not found or average doc size is '
-            'incorrect', self.coll)
-      return size
-
-  def _get_document_count(self):
+      name_space = '%s.%s' % (self.db, self.coll)
+      return (client[self.db].command(
+          'splitVector',
+          name_space,
+          keyPattern={'_id': 1},
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: my understanding is to use ascending index

File: sdks/python/apache_beam/io/mongodbio.py (hunk 23 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: yeah, that sounds better.

File: sdks/python/apache_beam/io/mongodbio.py (hunk 24 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio.py]: precommit tests covers python2.7, 3.5, 3.6, 3.7

File: sdks/python/apache_beam/io/mongodbio_test.py (hunk 26 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: Added comments to explain the assumption that every document is 1 Mb for simplicity of tests.

File: sdks/python/apache_beam/io/mongodbio_test.py (hunk 25 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: Done

File: sdks/python/apache_beam/io/mongodbio_test.py (hunk 27 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: done.

File: sdks/python/apache_beam/io/mongodbio_test.py (hunk 28 above)
Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: done.

**Comment by manuelaguilar** [comment]
@y1chi  I have patched my forked 2.14.0 with the fixes in this ticket. What would be a good performance test to verify this fix?

**Comment by y1chi** [comment]
> @y1chi I have patched my forked 2.14.0 with the fixes in this ticket. What would be a good performance test to verify this fix?

Probably can try with your own data set similar to https://github.com/apache/beam/blob/75150df8bb02cdafc983305631506d70d6039f86/sdks/python/apache_beam/io/mongodbio_it_test.py ?

**Comment by y1chi** [comment]
@jkff @chamikaramj updated the PR, would you mind give it another look?

File: sdks/python/apache_beam/io/mongodbio_test.py [hunk 30]
```diff
@@ -210,6 +323,74 @@ def test_write_to_mongodb_with_generated_id(self, mock_client):
...
+    stop_int = 2**96 - 1
+    start = _ObjectIdHelper.int_to_id(start_int)
+    stop = _ObjectIdHelper.int_to_id(stop_int)
+    test_cases = ([start_int, stop_int, 2**32, 2**32 - 1, 2**64, 2**64 - 1] +
+                  [random.randint(start_int, stop_int) for _ in range(100)])
+    tracker = _ObjectIdRangeTracker()
+    for pos in test_cases:
+      id = _ObjectIdHelper.int_to_id(pos - start_int)
+      desired_fraction = (pos - start_int) / (stop_int - start_int)
+      self.assertAlmostEqual(tracker.position_to_fraction(id, start, stop),
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/mongodbio_test.py]: Please run this test 10k times and make sure that this is not flaky.

Comment (y1chi) [reply sdks/python/apache_beam/io/mongodbio_test.py]: tested in 10k loop, all passed on my laptop.

**Comment by chamikaramj** [comment]
Eugene seems to be OOO. So I'll go ahead and merge this so that this can be cherry-picked to 2.15.0 release.
